### PR TITLE
feat(EMS-1027): no pdf - account sign in, creation - unverified email enhancements

### PR DESCRIPTION
**Pull request by ttbarnes**
## Introduction :pencil2:

This PR updates the account creation and sign in processes so that:

- If an account created, is unverified and a user submits the "create account" form with valid credentials a "confirm/verify email" email is sent.
- If an account created, is unverified and a user submits the "sign in" form with the valid credentials, a "confirm/verify email" email is sent.
- If a user successfully signs into their account, mark the account's `isInactive` status flag as `false`.

## Resolution :heavy_check_mark:

- Update `completeAndSubmitCreateAccountForm` cypress command to accept a custom `PASSWORD` value.
- Update error messages content strings to have:
  - `ACCOUNT_ALREADY_EXISTS`
  - `ACCOUNT_ALREADY_EXISTS_INCORRECT`
- Create a new E2E helper function, `createDateOneMonthInThePast`.
- Create E2E tests for submitting the "create an account" form with existing account data, with:
  - An unverified account, invalid password.
  - An unverified account, valid password.
  - A verified account, invalid password.
  - A verified account, valid password.
- Create E2E tests for submitting the "sign in" form with existing account data, with:
  - An expired verification token.
  - An expired verification token, account is marked as inactive.
  - An unverified account, no expired verification token.
- Rename API `accountChecks` helper function to `accountSignInChecks`.
- Update `accountSignInChecks` to _not_ check for verification token expiry. Instead, it now simply checks if an account is verified or not.
- Update the "create an account" GraphQL resolver to:
  - Not return `success=false` if an account already exists.
  - Check if the provided variables have valid credentials that match an existing account. If so and if the account is unverified, call `confirmEmailAddressEmail.send`.
  - Conditionally return some new flags for UI consumption: `alreadyExists`, `isVerified`.
- Update `generateOTPAndUpdateAccount` API helper to always mark an account's `isInActive` flag as false.
- Update the UI's `YOUR_DETAILS` POST controller to conditionally render different validation errors, depending on if the account is verified or not.
- Create/tweak `YOUR_DETAILS` validation rules:
  - `accountAlreadyExistsAlreadyVerifiedValidation`
  - `accountAlreadyExistsInvalidPasswordValidation`

## Miscellaneous :heavy_plus_sign:

- Create new `getAccountStatusById` helper function.
- Documentation, test description improvements.
- Various logging improvements.
- Update Keystone `Context` param documentation.

### CODE REVIEW

File: src/api/custom-resolvers/mutations/create-an-account/index.test.ts [hunk 1]
```diff
@@ -42,95 +47,322 @@ describe('custom-resolvers/create-an-account', () => {
     jest.resetAllMocks();
   });
 
-  beforeEach(async () => {
-    jest.resetAllMocks();
+  describe('when an account with the provided email does NOT already exist', () => {
```
Comment (ttbarnes) [inline src/api/custom-resolvers/mutations/create-an-account/index.test.ts:50]: Note: there's a lot of repeated `beforeAll` spys/setup in this file, would like to clean up after this PR. Will create a ticket.

File: e2e-tests/helpers/date.js [hunk 2]
```diff
@@ -19,3 +19,16 @@ export const createTimestampFromNumbers = (day, month, year) => new Date(`${mont
  * @returns {String} Formatted date
  */
 export const formatDate = (timestamp, dateFormat = DATE_FORMAT.DEFAULT) => format(new Date(timestamp), dateFormat);
+
+/**
+ * createDateOneMonthInThePast
+ * Create a date that is one month in the past.
+ * @returns {String} Date one month in the past
+ */
+export const createDateOneMonthInThePast = () => {
```
Comment (Zainzzkk) [inline e2e-tests/helpers/date.js:28]: just wondering if it is worth changing this to take a variable for any months in the past so we can use it in the future if needed for say 3 months etc

File: e2e-tests/insurance/cypress/e2e/journeys/account/sign-in/account-sign-in-with-an-unverified-account.spec.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      cy.getAccountByEmail(accountEmail).then((responseData) => {
+        const [firstAccount] = responseData;
+        account = firstAccount;
+      });
+    });
+
+    it(`should redirect to ${CONFIRM_EMAIL_RESENT}`, () => {
+      cy.navigateToUrl(signInUrl);
+
+      cy.completeAndSubmitSignInAccountForm({ assertRedirectUrl: false });
```
Comment (Zainzzkk) [inline e2e-tests/insurance/cypress/e2e/journeys/account/sign-in/account-sign-in-with-an-unverified-account.spec.js:57]: worth having a shared command for this as repeated 3 times?

File: src/api/custom-resolvers/mutations/account-sign-in/account-sign-in-checks/index.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ * If not, this means that the user has not verified their email address. We can therefore:
+ * 1) Reset the account's verification expiry.
+ * 2) Re-send the verification link email that was sent during account creation.
+ *
+ * Otherwise, the account is verified. We can therefore:
+ * 1) Generate an OTP
+ * 2) Update the account
+ * 3) Send an email with the OTP
+ *
+ * @param {Object} KeystoneJS context API
```
Comment (Zainzzkk) [inline src/api/custom-resolvers/mutations/account-sign-in/account-sign-in-checks/index.ts:19]: Context? for param

File: src/api/helpers/get-account-status-by-id/index.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { Account, Context } from '../../types';
+
+/**
+ * getAccountStatusById
+ * Get an account status by account status ID
+ * @param {Object} context: KeystoneJS context API
```
Comment (Zainzzkk) [inline src/api/helpers/get-account-status-by-id/index.ts:6]: Context?

File: src/api/custom-resolvers/mutations/account-sign-in/account-sign-in-checks/index.ts (hunk 4 above)
Comment (ttbarnes) [reply src/api/custom-resolvers/mutations/account-sign-in/account-sign-in-checks/index.ts:19]: Updated all instances of this in the API 👍

File: e2e-tests/insurance/cypress/e2e/journeys/account/sign-in/account-sign-in-with-an-unverified-account.spec.js (hunk 3 above)
Comment (ttbarnes) [reply e2e-tests/insurance/cypress/e2e/journeys/account/sign-in/account-sign-in-with-an-unverified-account.spec.js:57]: Agree, can we do this in a future PR, will create a ticket.

Will be lots of places to update this and we're running short on QA time before the PEN test

File: e2e-tests/helpers/date.js (hunk 2 above)
Comment (ttbarnes) [reply e2e-tests/helpers/date.js:28]: Good idea, could change this in the future if required? Happy to change if a blocker though.

**Review by Zainzzkk — APPROVED** [review]

File: e2e-tests/insurance/cypress/e2e/journeys/account/sign-in/account-sign-in-with-an-unverified-account.spec.js (hunk 3 above)
Comment (ttbarnes) [reply e2e-tests/insurance/cypress/e2e/journeys/account/sign-in/account-sign-in-with-an-unverified-account.spec.js:57]: Created EMS-3464

**Pull request merged**